### ISSUE TITLE
Issues with EasySIMBL and OS X 10.11 El Capitan

### ISSUE THREAD

**Issue by antons** [norio-nomura/EasySIMBL#26]
It appears that 10.11b1 is not affected by the issue described in #25.

However, [one of my plugins](https://github.com/antons/Gitbox-Tweaks) which previously injected into the target app just fine, now has issues.

When I restart the system, and launch the app for the first time, the plugin (and other plugins) are injected into the app. If I quit the app, and launch it again, I get the following console message.

```
09.06.15 18:06:32,626 Gitbox[1219]: Error loading /Users/Anton/Library/ScriptingAdditions/EasySIMBL.osax/Contents/MacOS/EasySIMBL:  dlopen(/Users/Anton/Library/ScriptingAdditions/EasySIMBL.osax/Contents/MacOS/EasySIMBL, 262): no suitable image found.  Did find:
	/Users/Anton/Library/ScriptingAdditions/EasySIMBL.osax/Contents/MacOS/EasySIMBL: open() failed with errno=24
```

**Comment by norio-nomura** [comment]
I have not yet tested on OS X 10.11, but I think it will be checked by attaching debugger to Gitbox.
On debugger's console, I did following:
```
(lldb) call (void*)dlopen("/Users/norio/Library/ScriptingAdditions/EasySIMBL.osax/Contents/MacOS/EasySIMBL",0)
(void *) $0 = 0x0000608000167980
```
and did resume the process.
On OS X 10.10.4 (14E33b), I got Gitbox Tweaks 1.2 working on Gitbox 1.6.2 with above.

Maybe it will be possible for `open()` or other APIs.

**Comment by jensfrancis** [comment]
Tested SafariStrand on EasySIMBL with OS X 10.11 Build 15A178w.  No dice.

**Comment by catlan** [comment]
Runtime protections
 
Injecting code into a process is equivalent to modifying the binary on disk
Processes are marked restricted by the kernel
- Main executable is protected on disk
- (other stuff went by fast, watch the video)

https://forums.developer.apple.com/message/7363#7363

**Comment by antons** [comment]
@catlan I’m not sure how rutime protection is related to my question, considering that the issue is only with one app, which _loads plugins_, but once, and only fails to do it again unless the system is restarted. Other plugins are injected into other apps without any issues.

**Comment by orbitly** [comment]
Can confirm it continues to not work on beta 2 (15A204h)

**Comment by d235j** [comment]
@orbitly: 10.11 developer preview 2 has begun enforcing code injection restrictions. You will need to boot into recovery mode and disable System Integrity Protection for EasySIMBL to work.

**Comment by catlan** [comment]
@d235j can you confirm that System Integrity Protection protects all apps from code injection or does it just protect system apps?

**Comment by fjolnir** [comment]
Disabling SIP didn't fix things for me. (TotalFinder on the other hand did start working after disabling it)

**Comment by ryanp-me** [comment]
@catlan @d235j I don't have the beta installed, but the WWDC slides say System Integrity Protection will "Prevent runtime attachment and code injection into system binaries." It wouldn't surprise me if injection still works with normal apps.

**Comment by norio-nomura** [comment]
Maybe #25 will prevent EasySIMBL working if SIP is disabled on OS X 10.11.

**Comment by w0lfschild** [comment]
Yup even with SIP off on 10.11 (15A204h) I'm not having success with anything loading.

**Comment by ryanp-me** [comment]
@norio-nomura It looks like that's probably the problem.

I just installed a VM of OS X 10.11. To ensure that SIP was enabled, I tried to create a file in `/System/` as `root` and I received an `operation not permitted` error. When I tried to inject code using my own method, it worked fine on some executables but not others.

    $ ls
    2015-06-28 22:34:10.827 ls[404:6219] TestInject Loaded
    Desktop
    Documents
    Downloads
    Library
    Movies
    Music
    Pictures
    Public

    $ /Applications/TextEdit.app/Contents/MacOS/TextEdit
    2015-06-28 22:36:14.378 TextEdit[405:6555] TestInject Loaded

I was able to inject code into Console, Terminal, TextEdit, ls, and a few others. However, I was not able to inject code into Finder, Safari, or Notes.

**Comment by norio-nomura** [comment]
FYI about SIP http://blog.binaryage.com/el-capitan-update/

**Comment by norio-nomura** [comment]
I confirmed that original SIMBL-0.9.9 placed at `/System/Library/ScriptingAdditions/` works on SIP enabled OS X 10.11 (15A204h).

Steps of installing SIMBL-0.9.9:
1. Disable SIP by `sudo nvram boot-args="rootless=0"` and reboot.
2. Run following on terminal and reboot.
```sh
sudo installer -verbose -pkg Downloads/SIMBL-0.9.9.pkg -target /
sudo rm -rf /System/Library/ScriptingAdditions/SIMBL.osax
sudo mv /Library/ScriptingAdditions/SIMBL.osax /System/Library/ScriptingAdditions/
sudo cp -p /System/Library/ScriptingAdditions/SIMBL.osax/Contents/Resources/SIMBL\ Agent.app/Contents/Resources/net.culater.SIMBL.Agent.plist /System/Library/LaunchAgents/
sudo sed -e "s/Library/System\/Library/" -i "" /System/Library/LaunchAgents/net.culater.SIMBL.Agent.plist
sudo nvram -d boot-args # enable SIP
# sudo reboot
```

After above steps, 
Plugins placed at `/Library/Application Support/SIMBL/Plugins` will be loaded by SIMBL.

[SafariStand](https://github.com/hetima/SafariStand/) built by myself of https://github.com/hetima/SafariStand/commit/53b1e5bd57ce39e5de6aee9bf330528f6106b438 works on Safari 9.0 (11601.1.35.2) with above setup.

Notes:
- Original `SIMBL Agent.app` does not support injecting into process that launched before `SIMBL Agent.app`. Verify `SIMBL Agent.app` is running before launching target app.
- If plugins does not work, check opened files of target app process using `Activity Monitor.app`. If plugin is listed in the opened files, plugin will need updating to support OS X 10.11. (Yes, SafariStand needed.)
- `SIMBL.osax` does not require codesign.
- plugins does not require codesign.
- plugins does not require owned by root.
- `/System/Library/LaunchAgents/net.culater.SIMBL.Agent.plist`' requires owned by root.

**Comment by norio-nomura** [comment]
I wrote a note about injection mechanism of EasySIMBL extended from SIMBL-0.9.9
https://gist.github.com/norio-nomura/f4346e9cb9db0a411779

**Comment by dvcrn** [comment]
@norio-nomura thanks for the instructions. I was able to compile and run stand with the setup above!

**Comment by Fofer** [comment]
norio-nomura, 

how does that demonstrate "SIMBL-0.9.9 working on SIP enabled OS X 10.11" if your very first step is "disabling SIP and rebooting?"

**Comment by w0lfschild** [comment]
@Fofer 

Because the last step is turning it back on... You only need to turn off SIP to install SIMBL.

**Comment by Fofer** [comment]
Ah, gotcha, thanks.

I really hope we'll be able to keep using SafariStand in OS X 10.11.

**Comment by norio-nomura** [comment]
I updated the comment.

**Comment by d235j** [comment]
For whatever it's worth, Apple has already stated that changing SIP via boot-args will not be supported in the release version of El Capitan.

**Comment by ryanp-me** [comment]
@d235j That's fine, as long as it can still be toggled from the recovery partition.

**Comment by antons** [comment]
@norio-nomura Have you ever encountered an issue where a plugin is injected when [the app is launched from Xcode](https://cloud.githubusercontent.com/assets/129705/8636875/7b1c8dd2-2880-11e5-81e1-3fa98447196d.png), but not when when launched from Finder? Using original SIMBL with your instructions.

**Comment by norio-nomura** [comment]
@antons Enabling debug logging may help you.
```sh
defaults write net.culater.SIMBL SIMBLLogLevel -int 0
```
from https://code.google.com/p/simbl/wiki/Tutorial

**Comment by antons** [comment]
@norio-nomura Thank you very much for continuing to check issues, even though you don’t use SIMBL yourself.

I’ll log this for anyone who may encounter the same issue. Unfortunately the logs weren’t helpful.

From Finder.
```
13.07.15 14:41:59,445 SIMBL Agent[506]: Gitbox started
13.07.15 14:41:59,446 SIMBL Agent[506]: app start notification: {
    NSApplicationBundleIdentifier = "com.oleganza.gitbox";
    NSApplicationName = Gitbox;
    NSApplicationPath = "/Applications/Gitbox.app";
    NSApplicationProcessIdentifier = 72631;
    NSApplicationProcessSerialNumberHigh = 0;
    NSApplicationProcessSerialNumberLow = 6231537;
    NSWorkspaceApplicationKey = <NSRunningApplication: 0x40013a820 (com.oleganza.gitbox - 72631)>;
}
13.07.15 14:41:59,446 SIMBL Agent[506]: checking bundle /Users/Anton/Library/Application Support/SIMBL/Plugins/Gitbox Tweaks SIMBL Plugin.bundle
13.07.15 14:41:59,447 SIMBL Agent[506]: checking target identifier com.oleganza.gitbox
13.07.15 14:41:59,447 SIMBL Agent[506]: send inject event
13.07.15 14:41:59,448 SIMBL Agent[506]: warning: failed to get scripting definition from /Applications/Gitbox.app; it may not be scriptable.
13.07.15 14:41:59,702 SIMBL Agent[506]: eventDidFail:'tvea' error:Error Domain=NSOSStatusErrorDomain Code=-1708 "The operation couldn’t be completed. (OSStatus error -1708.)" (errAEEventNotHandled: the AppleEvent was not handled by any handler ) UserInfo=0x400182700 {ErrorNumber=-1708} userInfo:{
    ErrorNumber = -1708;
}
```

From Xcode.
```
13.07.15 14:43:55,518 SIMBL Agent[506]: Gitbox started
13.07.15 14:43:55,519 SIMBL Agent[506]: app start notification: {
    NSApplicationBundleIdentifier = "com.oleganza.gitbox";
    NSApplicationName = Gitbox;
    NSApplicationPath = "/Applications/Gitbox.app";
    NSApplicationProcessIdentifier = 73651;
    NSApplicationProcessSerialNumberHigh = 0;
    NSApplicationProcessSerialNumberLow = 6247925;
    NSWorkspaceApplicationKey = <NSRunningApplication: 0x4001cd400 (com.oleganza.gitbox - 73651)>;
}
13.07.15 14:43:55,519 SIMBL Agent[506]: checking bundle /Users/Anton/Library/Application Support/SIMBL/Plugins/Gitbox Tweaks SIMBL Plugin.bundle
13.07.15 14:43:55,520 SIMBL Agent[506]: checking target identifier com.oleganza.gitbox
13.07.15 14:43:55,520 SIMBL Agent[506]: send inject event
13.07.15 14:43:55,521 SIMBL Agent[506]: warning: failed to get scripting definition from /Applications/Gitbox.app; it may not be scriptable.
13.07.15 14:43:55,974 Gitbox[73651]: Performance: Please update this scripting addition to supply a value for ThreadSafe for each event handler: "/System/Library/ScriptingAdditions/SIMBL.osax"
13.07.15 14:43:55,976 SIMBL Agent[506]: eventDidFail:'tvea' error:Error Domain=NSOSStatusErrorDomain Code=-1708 "The operation couldn’t be completed. (OSStatus error -1708.)" (errAEEventNotHandled: the AppleEvent was not handled by any handler ) UserInfo=0x400123900 {ErrorNumber=-1708} userInfo:{
    ErrorNumber = -1708;
}
13.07.15 14:43:56,084 Gitbox[73651]: load SIMBL plugins
13.07.15 14:43:56,084 Gitbox[73651]: SIMBL loaded by path /Applications/Gitbox.app <com.oleganza.gitbox>
13.07.15 14:43:56,084 Gitbox[73651]: checking bundle /Users/Anton/Library/Application Support/SIMBL/Plugins/Gitbox Tweaks SIMBL Plugin.bundle
13.07.15 14:43:56,193 Gitbox[73651]: loaded /Users/Anton/Library/Application Support/SIMBL/Plugins/Gitbox Tweaks SIMBL Plugin.bundle
```

**Comment by alisookasa** [comment]
@antons I got this error once when I tried to run SIMBL Agent app without running through the pkg installer; the simbl.pkg register the ScritptingAdditions under /Library, which is essential to make the injection work.

In my case, I had to run the [installer](http://www.culater.net/dl/files/SIMBL-0.9.9.zip) once, then re-run the SIMBL Agent app.

**Comment by norio-nomura** [comment]
@antons It seems `SIMBL Agent.app` recognized launching of target application, but failed injecting `SIMBL.osax` into target application. If you could inject by sending Apple Event manually, The issue may be caused by `SIMBL Agent.app`.
EasySIMBL's timing of sending Apple Event is slightly different from original SIMBL. It caused from change of observing launching applications method for fixing injection issues while I found on creating my plugins.
I don't remember the detail of those issues, but it might had contained one of debugging issue.

**Comment by antons** [comment]
My issue (Gitbox Tweaks not loading into Gitbox, while other plugins work) is gone after installing 10.11.0 beta 4.

**Comment by kennysmoothx** [comment]
@norio-nomura Your steps helped me get SIMBL working again.

Thanks!

**Comment by levifig** [comment]
Disabling SIP to install SIMBL and re-enabling it afterwards seems to do the trick. Thank you guys. :)

**Comment by KenV54** [comment]
I've got Afloat, which depends upon SIMBL, working on most apps except for Safari in 10.11 public preview beta 15A262e. I just manually added Afloat to the plugins in the SIMBL directory in the library. I still cannot get it to work with Safari, which is the way it used to be with Safari, at least before Yosemite, as I recall. I haven't yet done the SIP fix, because I'm thinking it might then have to be done again every time the OS X is updated. I'll try a reboot and comment back here if it works. If no comment, it didn't work of Safari.

**Comment by KenV54** [comment]
Well, I'll comment anyway. It didn't work in Safari. I wonder if Apple has changed Safari back such that Afloat doesn't work in Safari no matter what, just as it used to be. Afloat works fine for me in Chrome and many other apps under El Capitan.

**Comment by bwiernik** [comment]
In preview beta 15A278b, the steps described no longer work. After installing SIMBL, the lines of the script that modify /System/Library/ScriptingAdditions/SIMBL.osax fail, even after disabling SIP.

**Comment by KenV54** [comment]
I have the same preview beta 15A278b, have taken no steps to modify SIMBL, and, in fact, have no SIMBL.osax in the directory mentioned. My app that depends upon SIMBL, Afloat, works for most things except for Safari, as I mentioned in the post above. Prior to EC, Afloat was working for Safari, too. I wonder if Safari was changed back in some way to what it used to be in Mavericks, in which Afloat wouldn't work in Safari but did work in Chrome and Firefox.

**Comment by norio-nomura** [comment]
@bwiernik @KenV54 configuring SIP has been changed.
I updated https://github.com/norio-nomura/EasySIMBL/issues/26#issuecomment-117028426 for OS X 10.11 Developer Beta 8 (15A279b).

**Comment by KenV54** [comment]
Amazing. It worked! Thank you! I now have Afloat working again in Safari! It had been working with other apps before I went through the SIP configuration, but not ink Safari, and now it does again.

I wonder if it will be necessary to go through this procedure every time I upgrade the El Capitan Preview beta or the official 10.11 release?

**Comment by BrysovVadik** [comment]
Safari 9.0 (11601.1.56). Did everything as described, but the icons appeared only in SideBar, and the tabs icon does not appear. In the Stand settings, Show Icon On Tab Bar checked. What can i do? Thanks.

**Comment by KenV54** [comment]
I have Afloat working with Safari 9 again. Haven't tried SafariStand. Afloat was working with the latest El Capitan public preview release, but when I did the install of the EC GM Candidate yesterday--it was a full install, no upgrade available--I had to go through the whole process again with the CSRUTIL Disable described above. 

I wonder if every OS upgrade in EC from now on will necessitate all those steps to make SIMBL work?

**Comment by BrysovVadik** [comment]
I did the procedure twice. No result.

**Comment by norio-nomura** [comment]
@BrysovVadik
>  In the Stand settings, Show Icon On Tab Bar checked.

It seems that SIMBL has successfully injected plugins into Safari.
The rest of issues should be filed to plugin's project.

**Comment by ballo** [comment]
Someone has already released a simple tool to disable SIP:

http://1i0ny.de/apps/rootlesstoggler/

**Comment by ryanp-me** [comment]
@ballo That only works on the first few betas, since it toggles rootless on and off by updating the nvram `boot-args` variable. In order to disable SIP on the more recent betas, you'll need to boot into recovery mode and use the `csrutil` command line utility.

**Comment by iamdavid99** [comment]
Works for me on the GM

**Comment by mdaddy** [comment]
which version of easysimbl did you install? 1.7.1 or 0.9.9?

(I am on El Capitan)

**Comment by mdaddy** [comment]
Actually, I got safarstand to work by following the steps here: https://github.com/norio-nomura/EasySIMBL/issues/26#issuecomment-117028426

However, when I right click on an image in safari, safari just crashes. I have tried it multiple times.

**Comment by InfAspire** [comment]
Trying with the release version of El Capitan, following the instructions above, and I can't get it to work. Tried twice. I do not see the SafariStand plugin loaded in the SIMBL Agent when I check the open files in activity monitor. The latest SafariStand is in both /Library/Application Support/SIMBL/Plugins and ~/Library/Application Support/SIMBL/Plugins. I was user owned in /Library/Application Support/SIMBL/Plugins and I tried also with root owned but neither worked. I tried installing SIMBL both from command line and from the package installer.

Has anybody gotten it working with the 10.11 release version?

**Comment by stewacide** [comment]
Afloat has worked for me (apart from Safari and other system apps) in the El Capitan betas. What versions of Afloat+SIMBL do I want to install to have it working with Safari as well? (apparently I need to do a full OSX install to upgrade out of the beta, which I'm assuming will break everything)

**Comment by KenV54** [comment]
@stewaside, you're correct about Afloat and everything but Safari with EC, even the latest versions. The only way to make Afloat work in Safari is to use the method in comment #26 above. It takes a little time, you have to go into recovery mode a couple times, but it works, and it seems to hold even with OS updates. It probably won't hold with an EC fresh install, and it's unclear at this point whether people with the OS 10.11.1 public beta will have to do a fresh install of 10.11 now that it's been released, or whether we can jus upgrade over that.

**Comment by norio-nomura** [comment]
I updated https://github.com/norio-nomura/EasySIMBL/issues/26#issuecomment-117028426 for OS X 10.11 (15A284).

**Comment by sha256** [comment]
@norio-nomura 
`sudo sed -e "s/Library/System\/Library/" -i "" /System/Library/LaunchAgents/net.culater.SIMBL.Agent.plist`

This gives me `sed: can't read : No such file or directory`

**Comment by sha256** [comment]
I got it the SIMBL working by #26 (comment). Safari Stand also shows up in the menu bar of Safari. But Favicon in the tab is not working.

<img width="505" alt="screen shot 2015-10-02 at 10 56 17 am" src="https://cloud.githubusercontent.com/assets/1493554/10239774/880fba1e-68f4-11e5-94cf-f0a592b7fccc.png">

**Comment by levifig** [comment]
I really miss those tab favicons… :(

**Comment by repman** [comment]
I really miss those Safari Cookies...

**Comment by proteins247** [comment]
I'll chime in to say that I followed https://github.com/norio-nomura/EasySIMBL/issues/26#issuecomment-117028426 to successfully install Stand. But no favicons on tabs and Safari crashed when I right-clicked on a hyperlink.

**Comment by Sufl** [comment]
I successfully installed SIMBL on OS X 10.11.1 via Terminal as described in https://github.com/norio-nomura/EasySIMBL/issues/26#issuecomment-117028426 but it did not work.  I noticed that there was no plugin in Library/SIMBL/Plugins.  I tried it again but still no plugin there.  I then tried to install SarariStand but that did not seem to work.

I hate the grey sidebar and desperately want the colorful on back.  Any advice?

**Comment by swrobel** [comment]
@Sufl you need to manually copy the SafariStand file into `/Library/Application Support/SIMBL/Plugins` (or maybe `~/Library/Application Support/SIMBL/Plugins`, I haven't tried that path yet).

**Comment by Sufl** [comment]
Many thanks swrobel.  Eventually I got Stand in my Safari but it is the Finder Sidebar colors that I am desperate to get back.  I've spent 2 days trying.  I thought maybe having SafariStand might have helped me to achieve that but it didn't unless there is something else I am supposed to do.  I don't know what else to try.  Any help with this would be very much appreciated.

**Comment by iamdavid99** [comment]
Try XtraFinder

**Comment by Fofer** [comment]
Sufl, SafariStand has nothing to do with the Finder Sidebar colors.  That is handled by another SIMBL plug-in.  And there are a few out there that have enabled this in the past.  Which one are you using?  And, are you sure it's even compatible with OS X 10.11?  Probably not.  XtraFinder is.  Try that.

**Comment by swrobel** [comment]
@Sufl +1 for [XtraFinder](https://www.trankynam.com/xtrafinder/) - have not tried Finder Sidebar colors but they're a supported option. Apologies to everyone else for getting so off-topic here.

**Comment by Sufl** [comment]
Thank you everyone for your help.  I thought https://github.com/norio-nomura/EasySIMBL/issues/26#issuecomment-117028426 was what you had to do to get the colorful Finder sidebar back in El Capitan as this post gave the solution to that for Yosemite.

Yes I successfully got the colored Finder sidebar back using XtraFinder but it only works with SIP disabled and I thought it might be best to have it enabled.  I have spent days searching for other ways to get the color back in Finder and trying various things but no joy which was why I was asking for help here as it helped me before.  I had thought installing SIMBL, as described above, was the answer but it seems not.

If there is any possibility that Finder's color can be returned then I would be most keen to hear it.  Thanks again.

**Comment by levifig** [comment]
Sufl, the return of the color icons is unlikely. That was a deliberate change by the Finder and OS X designers. I (and many people) actually like the change but can understand that some people don't. Having said that, color icons on the Finder's sidebar, for the foreseeable future, will have to rely on hacks and workarounds, and SIP will always prevent them. As such, it is up to you to choose which you want more: color icons or SIP. :)

Cheers!

**Comment by w0lfschild** [comment]
@levifig @Sufl 

You can use XtraFinder with SIP on. https://www.reddit.com/r/OSXTweaks/comments/3o3h6k/question_is_there_actual_demand_for_more_native/cvuxvtq

**Comment by Sufl** [comment]
Dear w0lfschild, you did it for me and I think you are a miracle worker and wonderful.  Your answer did the trick for me and, unbelievably, I have got my colorful Finder sidebar back - even after closing down, unlike the Yosemite fix I got earlier on here.  So I am absolutely delighted because I had latterly given up.  That trick was magic and I thank you so much for your help.  You have made me VERY happy :-)

**Comment by mdaddy** [comment]
Just tried XtraFinder Unfortunately, doesn't work after you turn SIP back on. 

1. Download ExtraFinder
2. Reboot —> Disable SIP —> Reboot
3. Install XtraFinder. Works like a charm. All colorful sidebar and labels are back. 
4. Reboot —> Enable back SIP —> Reboot
Now XtraFinder doesn't even launch.

So XtraFinder works if you are okay with keeping SIP disabled. 

@Sufl - I am guessing you kept SIP off for XtraFinder to keep working?

**Comment by mdaddy** [comment]
PS @w0lfschild  — I use cDock2. good stuff. thanks for that.

**Comment by Sufl** [comment]
Yes that was the problem for me too, mdaddy, till I tried w0lfschild solution, above, which allowed me to have colorful Finder sidebar icons with SIP ENABLED via XtraFinder!  I still can't believe it. I keep clicking on Finder to see if it is still true.  And it is.  I had tried cDock2 too but there was no option in it for Finder colorful sidebar icons in El Capitan which was what I had wanted.  But now I have that with the solution just mentioned - with SIP enabled.  So now I've got the best of both worlds.

**Comment by mdaddy** [comment]
boom. that was it. I have XtrafFinder running, with SIP on. 

thanks @Sufl  and @w0lfschild

**Comment by kenanthebarbarian** [comment]
Let me preface this by saying that I know nothing about any of this stuff, but I am generally good at following directions.

Anyway, I recently upgraded to OS X 10.11.3 and am trying to get Afloat to work again in Safari for the fairly pedestrian purpose of keeping a Twitch.tv chat window visible in overlay while playing computer games in fullscreen. 

First, I deleted all previous SIMBL and Afloat files and followed @norio-nomura's instructions posted June 30, 2015, to reinstall SIMBL-0.9.9.

Next, I reinstalled Afloat from [here](https://github.com/w0lfschild/SIMBLPlugins/tree/master/Afloat) and later followed @w0lfschild's directions posted Oct. 24, 2015, replacing "XtraFinder.osax" with "SIMBL.osax" where applicable.

I also followed @w0lfschild's directions from [here](https://github.com/w0lfschild/SIMBLPlugins), replacing \"NotificationCenter\" with \"Safari\" and "Snow Leopard" with "El Capitan."

End result: Afloat works in Safari only if I boot up with SIP disabled. (Which, I gather, is not preferable to having SIP _en_abled.)

Any ideas? (Preferably simple and clearly-explained ones?)

**Comment by cafink** [comment]
Using @norio-nomura's instructions from June 30th, I was able to get Afloat up and running in El Capitan.  I disabled SIP, installed SIMBL and Afloat, then re-enabled SIP, and everything initially appeared to work fine.

However, Afloat's options now randomly (as far as I can tell) disappear from the `Window` menu after a while.  The application (Sublime Text, in my case) remains transparent (my main reason for using Afloat), but the transparency can not be adjusted any further.  The Afloat options sometimes re-appear after closing and re-opening the application, but this usually requires several attempts.

What could be the cause of this behavior?

**Comment by octalmage** [comment]
I'm having issues with this, no matter what I do I can't get a plugin injected without using Applescript to send the inject SIMBL event. Any ideas? 

Also, I need to check the use SIMBL box in EaseSIMBL. This seems different from the other posts I've read. 

I get this error in the console: 

```
5/21/16 4:58:04.523 PM com.github.norio-nomura.SIMBL-Agent[5063]: #EasySIMBL checking target identifier com.trankynam.aText
5/21/16 4:58:04.523 PM com.github.norio-nomura.SIMBL-Agent[5063]: #EasySIMBL should install plugin /Users/jason.stallings/Library/Application Support/SIMBL/Plugins/bText.bundle
5/21/16 4:58:04.523 PM com.github.norio-nomura.SIMBL-Agent[5063]: #EasySIMBL send inject event
5/21/16 4:58:04.524 PM com.github.norio-nomura.SIMBL-Agent[5063]: warning: failed to get scripting definition from /Applications/aText.app; it may not be scriptable.
```

**Comment by alexchandel** [comment]
It seems like one way around this without disabling SIP would be to write a SIMBL kernel extension that specifically enables SIMBL's functionality.

**Comment by w0lfschild** [comment]
@alexchandel Sorry but that doesn't exactly work. System Integrity Protection blocks unsigned kext. You'd still have to partly disable System Integrity Protection.

**Comment by norio-nomura** [comment]
AFAIK, A certificate for signing kext can't be get usual way. We need to submit request to apple by e-mail. I don't yet try that.
Developer ID certificate can't be used for signing kext as I tested before.

**Comment by w0lfschild** [comment]
@norio-nomura 

I wouldn't bother. Spyresoft tried this with DockMod and their signing was revoked (after being approved) and their kext was added to a kext blacklist that was silently pushed out to all Macs. Now Sypresoft and DockMod are abandonware as their site has gone down and all traces of them vanished. 

http://enterprisemac.bruienne.com/2016/03/04/kext-friends-forever/
https://www.spyresoft.com/dockmod/

**Comment by norio-nomura** [comment]
@w0lfschild Thanks for information. I didn't know about the incident of DockMod.

**Comment by alexchandel** [comment]
@w0lfschild right, a certificate would be required. But there are several open source kext projects with certificates (tun/tap and osxfuse come to mind).

As for DockMod, it seems these tweets:

> So it seems you can use DockMod to inject arbitrary signed library at least against Dock process :X
> 
> So it seems like that DockMod might be a nice free injection kernel driver for malware authors ;-) OOOOOOPPPPSSSSS ;-)

It seems Apple feared the existence of a kext that could inject arbitrary signed code into processes, or at least feared such a kext would be used for malware.

Hypothetically, I think a 3rd party kext for code injection that Apple would approve (and not later revoke) would be one that: limited what injected code could do to a finite list of tasks/features (e.g., adding new menu items to the menu bar, writing to /Library/Caches); was able to prevent injected code from doing anything other than what the kext permitted; could grant or deny a bundle's request to inject code on a per-feature basis, essentially running only parts of a bundle; only loaded injected code from SIP protected files; could move bundles into SIP-protection upon request; and required the user to explicitly tell it (i.e. choose in a SIP-protected, uninjectable GUI) to add a particular file, to explicitly mark (i.e. check in a SIP-protected, uninjectable GUI) which processes it could inject into, and explicitly mark (i.e. check in a SIP-protected, uninjectable GUI) which of the file's requested injection features are to be enabled for each chosen process.

This would cover most of the corner cases, and would be consistent with Apple's recent security philosophy of "no implicit permissions."

I'm not saying it's easy/possible to limit what injected code could do, but XPC services might be the way to go for that.

**Comment by alexandre-g** [comment]
Hey guys, Im trying SIMBL for the first time and all I'm seeing in console is this error
`SIMBL Agent[506]: warning: failed to get scripting definition from /Applications/Test.app; it may not be scriptable.`

Any ideas? I've turned debug logging on but nothing..
`defaults write net.culater.SIMBL SIMBLLogLevel -int 0`

**Comment by w0lfschild** [comment]
@alexandre-g consider using [mySIMBL](https://github.com/w0lfschild/mySIMBL). All other SIMBL variants including the original SIMBL do not work properly or at all on the latest releases of macOS.

**Comment by alexandre-g** [comment]
@w0lfschild I actually did end up finding mySIMBL last night, and was even going to leave a comment here, but didn't get to it. Great work by the way! Worked perfectly the first time.
